File: ios/brave-ios/Sources/Basarunaa/Utils/BodyPolygon.swift
// Copyright 2026 dev&din. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import CoreGraphics
import Foundation

/// Body polygon construction + rasterisation, port direct du POC
/// `private/extensions/basarunaa/src/utils/body_polygon.js`. Utilisé par
/// `PPLCNetClassifier` pour gris-outer les pixels hors silhouette avant
/// inférence (parité macOS `preprocessForClassification`).
enum BodyPolygon {

  struct Polygon {
    let points: [CGPoint]
    let isBodyShaped: Bool
  }

  struct Mask {
    let data: [UInt8]   // 0 ou 1, taille width * height
    let width: Int
    let height: Int
  }

  // Constantes POC (1:1).
  private static let kpConfidence: Double = 0.3
  private static let bodyWidthFactor: Double = 0.55
  private static let handExtend: Double = 0.3
  private static let footExtend: Double = 0.08
  private static let edgeSnap: Double = 0.05
  // Indexées par index COCO (0..16).
  private static let regionScale: [Double] = [
    0.8, 0.8, 0.8, 0.8, 0.8,   // head 0-4
    1.0, 1.0,                  // shoulders 5-6
    0.7, 0.7,                  // elbows 7-8
    0.6, 0.6,                  // wrists 9-10
    1.1, 1.1,                  // hips 11-12
    0.8, 0.8,                  // knees 13-14
    0.7, 0.7,                  // ankles 15-16
  ]

  /// Build a body-shaped polygon from 17 COCO keypoints (port direct
  /// `buildBodyPolygon`). Retourne un fallback bbox-rect si pas assez de
  /// keypoints fiables.
  static func buildPolygon(
    keypoints: [(point: CGPoint, confidence: Double)]?,
    bbox: CGRect,
    imageSize: CGSize
  ) -> Polygon {
    guard let kps = keypoints else { return bboxFallback(bbox) }

    let bx1 = Double(bbox.minX)
    let by1 = Double(bbox.minY)
    let bx2 = Double(bbox.maxX)
    let by2 = Double(bbox.maxY)
    let bw = bx2 - bx1
    let bh = by2 - by1

    let confident = kps.enumerated().filter { _, kp in kp.confidence >= kpConfidence }
    if confident.count < 4 { return bboxFallback(bbox) }

    // Half-width depuis les épaules (kp 5/6), fallback bbox * 0.25.
    let leftSh = kps.count > 5 ? kps[5] : nil
    let rightSh = kps.count > 6 ? kps[6] : nil
    var halfWidth: Double
    if let l = leftSh, let r = rightSh, l.confidence >= kpConfidence, r.confidence >= kpConfidence {
      halfWidth = abs(Double(r.point.x) - Double(l.point.x)) * bodyWidthFactor
    } else {
      halfWidth = bw * 0.25
    }
    halfWidth = max(halfWidth, bw * 0.2)

    var widened: [CGPoint] = []
    widened.reserveCapacity(confident.count * 2 + 8)
    for (idx, kp) in confident {
      let scale = idx < regionScale.count ? regionScale[idx] : 0.8
      let w = halfWidth * scale
      let kx = Double(kp.point.x)
      let ky = Double(kp.point.y)
      widened.append(CGPoint(x: kx - w, y: ky))
      widened.append(CGPoint(x: kx + w, y: ky))
    }

    // Head padding (au-dessus du keypoint le plus haut parmi 0-4).
    let headKps = (0..<min(5, kps.count)).filter { kps[$0].confidence >= kpConfidence }
    if !headKps.isEmpty {
      let topY = headKps.map { Double(kps[$0].point.y) }.min() ?? 0
      let sumX = headKps.reduce(0.0) { $0 + Double(kps[$1].point.x) }
      let headCx = sumX / Double(headKps.count)
      let headPadY = max(halfWidth * 0.6, bh * 0.08)
      let headPadX = max(halfWidth * 0.9, bw * 0.25)
      widened.append(CGPoint(x: headCx - headPadX, y: topY - headPadY))
      widened.append(CGPoint(x: headCx + headPadX, y: topY - headPadY))
    }

    // Extension main (coude 7→poignet 9, et 8→10).
    extendHand(kps: kps, elbowIdx: 7, wristIdx: 9, halfWidth: halfWidth, out: &widened)
    extendHand(kps: kps, elbowIdx: 8, wristIdx: 10, halfWidth: halfWidth, out: &widened)

    // Extension pieds (ankles 15/16 vers le bas).
    let footPad = bh * footExtend
    for ankleIdx in [15, 16] {
      guard ankleIdx < kps.count else { continue }
      let ankle = kps[ankleIdx]
      if ankle.confidence >= kpConfidence {
        let w = halfWidth * 0.7  // POC: REGION_SCALE.ankle = 0.7
        let ax = Double(ankle.point.x)
        let ay = Double(ankle.point.y)
        widened.append(CGPoint(x: ax - w, y: ay + footPad))
        widened.append(CGPoint(x: ax + w, y: ay + footPad))
      }
    }

    // Convex hull + scale to bbox bounds + edge snap.
    let hull = convexHull(widened)
    if hull.count < 3 { return bboxFallback(bbox) }

    var polyMinX = Double.infinity, polyMaxX = -Double.infinity
    var polyMinY = Double.infinity, polyMaxY = -Double.infinity
    for p in hull {
      polyMinX = min(polyMinX, Double(p.x))
      polyMaxX = max(polyMaxX, Double(p.x))
      polyMinY = min(polyMinY, Double(p.y))
      polyMaxY = max(polyMaxY, Double(p.y))
    }
    let polyCx = (polyMinX + polyMaxX) / 2
    let polyCy = (polyMinY + polyMaxY) / 2
    let bboxCx = (bx1 + bx2) / 2
    let bboxCy = (by1 + by2) / 2
    let sx = bw / max(polyMaxX - polyMinX, 1e-6)
    let sy = bh / max(polyMaxY - polyMinY, 1e-6)
    let scaled = hull.map { p -> CGPoint in
      CGPoint(
        x: bboxCx + (Double(p.x) - polyCx) * sx,
        y: bboxCy + (Double(p.y) - polyCy) * sy
      )
    }
    let snapped = snapToEdges(
      scaled,
      bbox: bbox,
      imageSize: imageSize,
      kps: kps
    )
    return Polygon(points: snapped, isBodyShaped: true)
  }

  /// Rasterize the polygon into a binary mask via scanline fill (port direct
  /// du POC `polygonToMask`). Mask values: 0 = outside, 1 = inside.
  static func polygonToMask(
    points: [CGPoint],
    imageWidth: Int,
    imageHeight: Int
  ) -> Mask {
    var data = [UInt8](repeating: 0, count: imageWidth * imageHeight)
    if points.count < 3 {
      return Mask(data: data, width: imageWidth, height: imageHeight)
    }

    var minY = Double(imageHeight), maxY = 0.0
    for p in points {
      if Double(p.y) < minY { minY = Double(p.y) }
      if Double(p.y) > maxY { maxY = Double(p.y) }
    }
    let yStart = max(0, Int(floor(minY)))
    let yEnd = min(imageHeight - 1, Int(ceil(maxY)))
    if yStart > yEnd {
      return Mask(data: data, width: imageWidth, height: imageHeight)
    }

    for y in yStart...yEnd {
      var inter: [Double] = []
      let yd = Double(y)
      for i in 0..<points.count {
        let a = points[i]
        let b = points[(i + 1) % points.count]
        let ay = Double(a.y), by = Double(b.y)
        if (ay <= yd && by > yd) || (by <= yd && ay > yd) {
          let denom = by - ay
          if abs(denom) > 1e-9 {
            inter.append(Double(a.x) + (yd - ay) / denom * Double(b.x - a.x))
          }
        }
      }
      inter.sort()
      var k = 0
      while k + 1 < inter.count {
        let x1 = max(0, Int(floor(inter[k])))
        let x2 = min(imageWidth - 1, Int(ceil(inter[k + 1])))
        if x1 <= x2 {
          for x in x1...x2 {
            data[y * imageWidth + x] = 1
          }
        }
        k += 2
      }
    }
    return Mask(data: data, width: imageWidth, height: imageHeight)
  }

  // MARK: - Internals

  private static func extendHand(
    kps: [(point: CGPoint, confidence: Double)],
    elbowIdx: Int,
    wristIdx: Int,
    halfWidth: Double,
    out: inout [CGPoint]
  ) {
    guard elbowIdx < kps.count, wristIdx < kps.count else { return }
    let elbow = kps[elbowIdx], wrist = kps[wristIdx]
    if elbow.confidence < kpConfidence || wrist.confidence < kpConfidence { return }
    let dx = Double(wrist.point.x - elbow.point.x)
    let dy = Double(wrist.point.y - elbow.point.y)
    let hx = Double(wrist.point.x) + dx * handExtend
    let hy = Double(wrist.point.y) + dy * handExtend
    let w = halfWidth * 0.6
    out.append(CGPoint(x: hx - w, y: hy))
    out.append(CGPoint(x: hx + w, y: hy))
  }

  private static func snapToEdges(
    _ points: [CGPoint],
    bbox: CGRect,
    imageSize: CGSize,
    kps: [(point: CGPoint, confidence: Double)]
  ) -> [CGPoint] {
    let imgW = Double(imageSize.width)
    let imgH = Double(imageSize.height)
    let by1 = Double(bbox.minY), by2 = Double(bbox.maxY)
    let bx1 = Double(bbox.minX), bx2 = Double(bbox.maxX)
    let hasAnkles = [15, 16].contains { i in
      i < kps.count && kps[i].confidence >= kpConfidence
    }
    let hasHead = [0, 1, 2].contains { i in
      i < kps.count && kps[i].confidence >= kpConfidence
    }
    let snapBottom = !hasAnkles && by2 / imgH > (1 - edgeSnap)
    let snapTop = !hasHead && by1 / imgH < edgeSnap
    let snapLeft = bx1 / imgW < edgeSnap
    let snapRight = bx2 / imgW > (1 - edgeSnap)
    if !snapBottom && !snapTop && !snapLeft && !snapRight { return points }

    var minX = Double.infinity, maxX = -Double.infinity
    var minY = Double.infinity, maxY = -Double.infinity
    for p in points {
      minX = min(minX, Double(p.x)); maxX = max(maxX, Double(p.x))
      minY = min(minY, Double(p.y)); maxY = max(maxY, Double(p.y))
    }
    let xR = max(maxX - minX, 1)
    let yR = max(maxY - minY, 1)
    let near = 0.15
    return points.map { p -> CGPoint in
      var x = Double(p.x), y = Double(p.y)
      if snapBottom && y > maxY - yR * near { y = imgH }
      if snapTop && y < minY + yR * near { y = 0 }
      if snapLeft && x < minX + xR * near { x = 0 }
      if snapRight && x > maxX - xR * near { x = imgW }
      return CGPoint(x: x, y: y)
    }
  }

  private static func bboxFallback(_ bbox: CGRect) -> Polygon {
    let pts: [CGPoint] = [
      CGPoint(x: bbox.minX, y: bbox.minY),
      CGPoint(x: bbox.maxX, y: bbox.minY),
      CGPoint(x: bbox.maxX, y: bbox.maxY),
      CGPoint(x: bbox.minX, y: bbox.maxY),
    ]
    return Polygon(points: pts, isBodyShaped: false)
  }

  /// Andrew's monotone chain convex hull (port direct du POC `_convexHull`).
  private static func convexHull(_ pts: [CGPoint]) -> [CGPoint] {
    if pts.count < 3 { return pts }
    let sorted = pts.sorted { lhs, rhs in
      lhs.x == rhs.x ? lhs.y < rhs.y : lhs.x < rhs.x
    }
    let n = sorted.count

    func cross(_ o: CGPoint, _ a: CGPoint, _ b: CGPoint) -> Double {
      Double((a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x))
    }

    var lower: [CGPoint] = []
    for i in 0..<n {
      while lower.count >= 2 && cross(lower[lower.count - 2], lower[lower.count - 1], sorted[i]) <= 0 {
        lower.removeLast()
      }
      lower.append(sorted[i])
    }
    var upper: [CGPoint] = []
    for i in stride(from: n - 1, through: 0, by: -1) {
      while upper.count >= 2 && cross(upper[upper.count - 2], upper[upper.count - 1], sorted[i]) <= 0 {
        upper.removeLast()
      }
      upper.append(sorted[i])
    }
    lower.removeLast()
    upper.removeLast()
    return lower + upper
  }
}

File: android/java/org/chromium/chrome/browser/basarunaa/BodyPolygon.java
/* Copyright (c) 2026 The Browther Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * you can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.basarunaa;

import android.graphics.PointF;

import org.chromium.build.annotations.NullMarked;
import org.chromium.build.annotations.Nullable;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import org.chromium.chrome.browser.basarunaa.BasarunaaTypes.Bbox;
import org.chromium.chrome.browser.basarunaa.BasarunaaTypes.Keypoint;

/**
 * Build a body-shaped polygon from COCO keypoints — port natif de
 * {@code private/extensions/basarunaa/src/utils/body_polygon.js} (= aussi
 * {@code src/core/body-polygon.ts}).
 *
 * <p>Utilisé pour grayer le background autour du body crop avant que
 * {@link PplcNetClassifier} ne classifie — sinon le classifier voit la
 * scène entière dans la bbox (foule, mobilier) ce qui dégrade gender accuracy.
 *
 * <p>Algorithme :
 * <ol>
 *   <li>Estime la largeur du corps depuis les épaules (kp 5, 6).</li>
 *   <li>Crée 2 points (gauche, droite) par keypoint visible, scalés par
 *       région (tête, épaules, hanche, etc.)</li>
 *   <li>Étend les mains au-delà des poignets (extrapolation coude→poignet).</li>
 *   <li>Étend les pieds sous les chevilles (8% bbox height).</li>
 *   <li>Construit le convex hull (Andrew monotone chain).</li>
 *   <li>Rescale le hull pour remplir la bbox.</li>
 *   <li>Snap aux bords image si la personne est croppée par les bords.</li>
 * </ol>
 */
@NullMarked
public final class BodyPolygon {
    private static final float KP_CONFIDENCE = 0.3f;
    private static final float BODY_WIDTH_FACTOR = 0.55f;
    private static final float HAND_EXTEND = 0.3f;
    private static final float FOOT_EXTEND = 0.08f;
    private static final float EDGE_SNAP_THRESHOLD = 0.05f;
    private static final float NEAR_FRAC = 0.15f;

    private static final float REGION_HEAD = 0.8f; // 0-4
    private static final float REGION_SHOULDER = 1.0f; // 5-6
    private static final float REGION_ELBOW = 0.7f; // 7-8
    private static final float REGION_WRIST = 0.6f; // 9-10
    private static final float REGION_HIP = 1.1f; // 11-12
    private static final float REGION_KNEE = 0.8f; // 13-14
    private static final float REGION_ANKLE = 0.7f; // 15-16

    private BodyPolygon() {}

    public static final class Result {
        public final List<PointF> points;
        public final boolean isBodyShaped;

        Result(List<PointF> points, boolean isBodyShaped) {
            this.points = points;
            this.isBodyShaped = isBodyShaped;
        }
    }

    /**
     * Construit le polygone autour du body. Si moins de 4 keypoints visibles,
     * fallback sur les 4 coins de la bbox.
     *
     * @param keypoints 17 COCO ou null
     * @param bbox bbox body en coords image source
     * @param imgW largeur image source
     * @param imgH hauteur image source
     */
    public static Result build(
            Keypoint @Nullable [] keypoints, Bbox bbox, int imgW, int imgH) {
        final float bw = bbox.width();
        final float bh = bbox.height();

        if (keypoints == null) return bboxFallback(bbox);

        int confidentCount = 0;
        for (Keypoint k : keypoints) {
            if (k != null && k.confidence >= KP_CONFIDENCE) confidentCount++;
        }
        if (confidentCount < 4) return bboxFallback(bbox);

        // Half-width depuis les épaules.
        final Keypoint lSh = keypoints[5];
        final Keypoint rSh = keypoints[6];
        float halfWidth;
        if (lSh.confidence >= KP_CONFIDENCE && rSh.confidence >= KP_CONFIDENCE) {
            halfWidth = Math.abs(rSh.x - lSh.x) * BODY_WIDTH_FACTOR;
        } else {
            halfWidth = bw * 0.25f;
        }
        halfWidth = Math.max(halfWidth, bw * 0.2f);

        final ArrayList<PointF> widened = new ArrayList<>();
        for (int i = 0; i < keypoints.length; i++) {
            final Keypoint k = keypoints[i];
            if (k.confidence < KP_CONFIDENCE) continue;
            final float scale = getRegionScale(i);
            final float w = halfWidth * scale;
            widened.add(new PointF(k.x - w, k.y));
            widened.add(new PointF(k.x + w, k.y));
        }

        // Head padding au-dessus des keypoints head visibles.
        float topY = Float.POSITIVE_INFINITY;
        float headCxSum = 0f;
        int headCount = 0;
        for (int i = 0; i <= 4; i++) {
            final Keypoint k = keypoints[i];
            if (k.confidence < KP_CONFIDENCE) continue;
            if (k.y < topY) topY = k.y;
            headCxSum += k.x;
            headCount++;
        }
        if (headCount > 0) {
            final float headCx = headCxSum / headCount;
            final float headPadY = Math.max(halfWidth * 0.6f, bh * 0.08f);
            final float headPadX = Math.max(halfWidth * 0.9f, bw * 0.25f);
            widened.add(new PointF(headCx - headPadX, topY - headPadY));
            widened.add(new PointF(headCx + headPadX, topY - headPadY));
        }

        // Hand extrapolation : coude→poignet → main extra.
        extendHand(keypoints, 7, 9, halfWidth, widened);
        extendHand(keypoints, 8, 10, halfWidth, widened);

        // Foot extension : ankle + footExtend en y.
        final float footExtend = bh * FOOT_EXTEND;
        for (int ankleIdx : new int[] {15, 16}) {
            final Keypoint ankle = keypoints[ankleIdx];
            if (ankle.confidence < KP_CONFIDENCE) continue;
            final float w = halfWidth * REGION_ANKLE;
            widened.add(new PointF(ankle.x - w, ankle.y + footExtend));
            widened.add(new PointF(ankle.x + w, ankle.y + footExtend));
        }

        // Convex hull.
        final List<PointF> hull = convexHull(widened);
        if (hull.size() < 3) return bboxFallback(bbox);

        // Rescale pour remplir la bbox.
        float polyMinX = Float.POSITIVE_INFINITY, polyMaxX = Float.NEGATIVE_INFINITY;
        float polyMinY = Float.POSITIVE_INFINITY, polyMaxY = Float.NEGATIVE_INFINITY;
        for (PointF p : hull) {
            if (p.x < polyMinX) polyMinX = p.x;
            if (p.x > polyMaxX) polyMaxX = p.x;
            if (p.y < polyMinY) polyMinY = p.y;
            if (p.y > polyMaxY) polyMaxY = p.y;
        }
        final float polyCx = (polyMinX + polyMaxX) / 2f;
        final float polyCy = (polyMinY + polyMaxY) / 2f;
        final float bboxCx = (bbox.x1 + bbox.x2) / 2f;
        final float bboxCy = (bbox.y1 + bbox.y2) / 2f;
        final float dx = polyMaxX - polyMinX;
        final float dy = polyMaxY - polyMinY;
        final float sx = bw / (dx == 0f ? 1f : dx);
        final float sy = bh / (dy == 0f ? 1f : dy);

        final ArrayList<PointF> scaled = new ArrayList<>(hull.size());
        for (PointF p : hull) {
            scaled.add(new PointF(
                    bboxCx + (p.x - polyCx) * sx,
                    bboxCy + (p.y - polyCy) * sy));
        }

        // Edge snapping si la personne touche les bords.
        final List<PointF> snapped = snapToEdges(scaled, bbox, imgW, imgH, keypoints);

        return new Result(snapped, true);
    }

    /**
     * Teste si un point (x, y) est à l'intérieur du polygone via raycast.
     * Utilisé par {@link PplcNetClassifier} pour grayer hors polygon.
     */
    public static boolean contains(List<PointF> polygon, float x, float y) {
        boolean inside = false;
        final int n = polygon.size();
        for (int i = 0, j = n - 1; i < n; j = i++) {
            final PointF a = polygon.get(i);
            final PointF b = polygon.get(j);
            if (((a.y > y) != (b.y > y))
                    && (x < (b.x - a.x) * (y - a.y) / (b.y - a.y) + a.x)) {
                inside = !inside;
            }
        }
        return inside;
    }

    private static float getRegionScale(int idx) {
        if (idx <= 4) return REGION_HEAD;
        if (idx <= 6) return REGION_SHOULDER;
        if (idx <= 8) return REGION_ELBOW;
        if (idx <= 10) return REGION_WRIST;
        if (idx <= 12) return REGION_HIP;
        if (idx <= 14) return REGION_KNEE;
        return REGION_ANKLE;
    }

    private static void extendHand(
            Keypoint[] kps, int elbowIdx, int wristIdx, float halfWidth,
            ArrayList<PointF> out) {
        final Keypoint elbow = kps[elbowIdx];
        final Keypoint wrist = kps[wristIdx];
        if (elbow.confidence < KP_CONFIDENCE || wrist.confidence < KP_CONFIDENCE) return;
        final float dxL = wrist.x - elbow.x;
        final float dyL = wrist.y - elbow.y;
        final float hx = wrist.x + dxL * HAND_EXTEND;
        final float hy = wrist.y + dyL * HAND_EXTEND;
        final float w = halfWidth * 0.6f;
        out.add(new PointF(hx - w, hy));
        out.add(new PointF(hx + w, hy));
    }

    private static List<PointF> snapToEdges(
            List<PointF> points, Bbox bbox, int imgW, int imgH, Keypoint[] keypoints) {
        boolean hasAnkles = false;
        boolean hasHead = false;
        for (int i : new int[] {15, 16}) {
            if (keypoints[i].confidence >= KP_CONFIDENCE) {
                hasAnkles = true;
                break;
            }
        }
        for (int i : new int[] {0, 1, 2}) {
            if (keypoints[i].confidence >= KP_CONFIDENCE) {
                hasHead = true;
                break;
            }
        }
        final boolean snapBottom = !hasAnkles && bbox.y2 / imgH > (1f - EDGE_SNAP_THRESHOLD);
        final boolean snapTop = !hasHead && bbox.y1 / imgH < EDGE_SNAP_THRESHOLD;
        final boolean snapLeft = bbox.x1 / imgW < EDGE_SNAP_THRESHOLD;
        final boolean snapRight = bbox.x2 / imgW > (1f - EDGE_SNAP_THRESHOLD);
        if (!snapBottom && !snapTop && !snapLeft && !snapRight) return points;

        float minX = Float.POSITIVE_INFINITY, maxX = Float.NEGATIVE_INFINITY;
        float minY = Float.POSITIVE_INFINITY, maxY = Float.NEGATIVE_INFINITY;
        for (PointF p : points) {
            if (p.x < minX) minX = p.x;
            if (p.x > maxX) maxX = p.x;
            if (p.y < minY) minY = p.y;
            if (p.y > maxY) maxY = p.y;
        }
        final float xRange = Math.max(1f, maxX - minX);
        final float yRange = Math.max(1f, maxY - minY);

        final ArrayList<PointF> out = new ArrayList<>(points.size());
        for (PointF p : points) {
            float x = p.x;
            float y = p.y;
            if (snapBottom && y > maxY - yRange * NEAR_FRAC) y = imgH;
            if (snapTop && y < minY + yRange * NEAR_FRAC) y = 0f;
            if (snapLeft && x < minX + xRange * NEAR_FRAC) x = 0f;
            if (snapRight && x > maxX - xRange * NEAR_FRAC) x = imgW;
            out.add(new PointF(x, y));
        }
        return out;
    }

    private static Result bboxFallback(Bbox bbox) {
        final ArrayList<PointF> pts = new ArrayList<>(4);
        pts.add(new PointF(bbox.x1, bbox.y1));
        pts.add(new PointF(bbox.x2, bbox.y1));
        pts.add(new PointF(bbox.x2, bbox.y2));
        pts.add(new PointF(bbox.x1, bbox.y2));
        return new Result(pts, false);
    }

    /** Andrew's monotone chain convex hull. */
    private static List<PointF> convexHull(List<PointF> points) {
        if (points.size() < 3) return new ArrayList<>(points);
        final ArrayList<PointF> sorted = new ArrayList<>(points);
        sorted.sort(Comparator.comparing((PointF p) -> p.x).thenComparing(p -> p.y));
        final int n = sorted.size();

        final ArrayList<PointF> lower = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            while (lower.size() >= 2
                    && cross(lower.get(lower.size() - 2),
                            lower.get(lower.size() - 1), sorted.get(i)) <= 0f) {
                lower.remove(lower.size() - 1);
            }
            lower.add(sorted.get(i));
        }
        final ArrayList<PointF> upper = new ArrayList<>();
        for (int i = n - 1; i >= 0; i--) {
            while (upper.size() >= 2
                    && cross(upper.get(upper.size() - 2),
                            upper.get(upper.size() - 1), sorted.get(i)) <= 0f) {
                upper.remove(upper.size() - 1);
            }
            upper.add(sorted.get(i));
        }
        lower.remove(lower.size() - 1);
        upper.remove(upper.size() - 1);
        lower.addAll(upper);
        return lower;
    }

    private static float cross(PointF o, PointF a, PointF b) {
        return (a.x - o.x) * (b.y - o.y) - (a.y - o.y) * (b.x - o.x);
    }
}
